### PR TITLE
LaTeX: Make sure sphinxVerbatim environment is not cut into pieces

### PR DESCRIPTION
**Pull request by mgeier**
#3022 breaks the LaTeX output of [nbsphinx](https://github.com/spatialaudio/nbsphinx) because I was assuming that the whole verbatim environment is contained in a single list element of `self.body`.

See https://github.com/spatialaudio/nbsphinx/issues/78.

This PR makes sure that only a single call to `self.body.append()` is used. The content (after concatenation) should be unchanged.

### CODE REVIEW

**Comment by jfbu** [comment]
LGTM from Sphinx point of view ! (with nits)

**Comment by jfbu** [comment]
@tk0miya could be merged in stable for 1.5.1 ?

**Review by jfbu — CHANGES_REQUESTED** [review]
would be better with single `+=`:
```
if self.table and not self.in_footnote:
    hlcode += '\\end{sphinxVerbatimintable}'
else:
    hlcode += '\\end{sphinxVerbatim}'
self.body.append('\n' + hlcode + '\n')
```

**Comment by tk0miya** [comment]
I feel a bit strange for this proposal.
I think `self.body` is not an opened and not an official interface of writers.
But this proposal modifies it to fit the form which nbsphinx expects.
Does this means that `self.body` is an interface?

I can agree merging this. But, on the other hand, I can't agree such an interface.
I believe nbsphinx should refer whole of `self.body`.

**Comment by mgeier** [comment]
Thanks @jfbu, I've changed the PR according to your suggestion.
It's still a bit strange that the "begin" tag is replaced with `replace()`, while the "end" tag is replaced with indexing and concatenation ...

@tk0miya I know that it's not an "official" interface, but hey, it's Python, everything is allowed!

Are you suggesting that I should do something like this in my node visitor?

```python
body_text = ''.join(self.body)
# replace/append stuff on body_text
self.body = [body_text]
```

This might work, but it doesn't look right, does it?

**Review by jfbu — APPROVED** [review]

**Comment by jfbu** [comment]
@tk0miya I agree that Sphinx developers should not get into the habit of supporting use by extensions of non official interface. In the case at hand however my commit at #3022 did

```
-            self.body.append('\n' + hlcode + '\\end{sphinxVerbatim%s}\n' %
-                             (self.table and 'intable' or ''))
+            self.body.append('\n' + hlcode + '\\end{sphinxVerbatim')
+            if self.table and not self.in_footnote:
+                self.body.append('intable')
+            self.body.append('}\n')
```

which did not make for very readable Python code (not to say that the former one was ;-). I probably did it that way in order to minimally change original code. (In fact I had even worse and you made a comment asking for improvement https://github.com/sphinx-doc/sphinx/pull/3022#discussion_r83348994). I feel now I should have done it from the start as implemented in this PR, hence I see this as opportunity to improve my Python contribution ...

**Comment by jfbu** [comment]
@mgeier replacement is always dangerous although it is highly improbable the string will occur in highlighted text returned from Pygments, precisely due to the fact that it escapes LaTeX.

Note: I see actually from Pygments ``formatters/latex.py`` (https://bitbucket.org/birkenfeld/pygments-main/src/a042025b350cd9c9461f7385d9ba0f13cdb01bb9/pygments/formatters/latex.py?at=default&fileviewer=file-view-default#latex.py-149) that it could be called asking in return for a specific environment name, hence the ``sphinxVerbatim`` could be obtained in one go, or ``sphinxVerbatimintable`` with no need of fiddling with `hlcode`.

**Comment by mgeier** [comment]
@jfbu Yeah I know `replace()` is potentially dangerous. That's why I wondered why you didn't change it when you changed the handling of the end tag.

And using the `envname` argument indeed looks like a much better way to do this!

Anyway, this shouldn't affect this PR, right?

And BTW, yes, it would be great to get this into Sphinx 1.5.1!

**Comment by jfbu** [comment]
But Pygments ``envname`` was added at version ``2.0``, which might be too recent for use now in Sphinx (I don't know). Anyway, that's option for future, I agree with merging this PR as is.

**Comment by shimizukawa** [comment]
> I know that it's not an "official" interface, but hey, it's Python, everything is allowed!

Because everything is allowed, we can change the way how latex writer hold the value in the self.body.... I feel it is a non-constructive argument.

**Comment by jfbu** [comment]
My feeling about this is that I 100% agree with @tk0miya @shimizukawa that Sphinx devs should indeed in no way at any time feel any sort of constraint on code which is not official interface. I feel in the specific case at hand  that the asked-for slight refactoring of a few lines in latex writer code makes for better readability and is not likely to cause other problems because it is closer to code from 1.4.x series. 

Sure it would have been better if problem with extension had been reported during beta release period (it is not the only one which missed that opportunity) and definitely extension must find a way without hacking `self.body.append()` which puts unduly constraint on Sphinx devs. (I don't know the specifics, perhaps a pure LaTeX approach on extension side is possible ; or not).

**Comment by mgeier** [comment]
@shimizukawa 

> Because everything is allowed, we can change the way how latex writer hold the value in the self.body....

Yes you can. And you did. And you were allowed to. But it would still have been more reasonable not to do it.

Anyway, I will try if it's possible to redefine the environment on the LaTeX side, so that I don't have to edit text that was previously appended to `self.body`.

But even if I find a way to do that, you should still merge this PR, because it just makes more sense, as @jfbu said.

**Comment by jfbu** [comment]
@mgeier please stay calm as you are not helping your cause. I am the one who changed the Python code, when I added to Sphinx latex the support of code-blocks in footnotes. In the long run it is definitely better if you find a way to achieve your goal without making assumptions like the fact that `self.boddy.append()` will be used only to inject latex environments in one go. I have not looked at what is at stake here, perhaps it is complicated or impossible on LaTeX side, perhaps it is only a matter of redefining `sphinxVerbatim` environment. Notice that it is now much easier than in the past to do these things, as formerly Sphinx simply overwrote original `fancyvrb`'s `Verbatim`. Such environments are very special in LaTeX and it is not easy to include them in other environments, etc..., but you can see how this is done in Sphinx 1.5 for Verbatim in table cells for example which removes the framing, and does other adaptations.

**Comment by tk0miya** [comment]
Okay. Refactoring is always welcomed. I'd like to merge this to 1.5.1.
Any comments?

**Comment by jfbu** [comment]
+1 for merge into 1.5.1

**Comment by tk0miya** [comment]
BTW, How about use `popbody()` and `pushbody()`of `LaTeXTranslator` class to obtain the generated code-block?
It replaces the `self.body` instance temporarily. So you can obtain whole of the code-block like following:
```
def visit_code_latex(self, node):
    self.pushbody([])
    ...

def depart_code_latex(self, node):
    code_block = ''.join(self.popbody())  # obtain generated code-block
    ...

    self.body.append(code_block)  # add customized code-block
```

It does not harm the sphinx-core and you can capture the generated source code.
I hope this helper method will help you.

Note: `popbody()` and `pushbody()` are provided only on LaTeX builder.

**Comment by tk0miya** [comment]
Merged!

**Comment by jfbu** [comment]
@tk0miya I am afraid this did not make it into 1.5.1 because it was merged into master only. It is my fault because I was aware this PR was on master. I should have requested explictely from @mgeier to rebase it on stable. But today was a busy day for me and I could not follow flow of events. I will merge manually into stable now so that it makes it into 1.5.2.

**Comment by jfbu** [comment]
Done at 108491d. Apologies to @tk0miya for not having helped more.

**Comment by tk0miya** [comment]
OMG, I'd also overlooked too.

@mgeier I'm very sorry. It will be released as 1.5.2. Please wait for the release.

**Comment by mgeier** [comment]
Thanks for merging!
I should have probably based this PR on `stable` in the first place, but I never know which branch to base PRs on. I'm normally confused by the available branches.
Is there some documentation that explains that?
And what does `stable` even mean for Sphinx?

@tk0miya Thanks for the hint with `pushbody()` and `popbody()`, they sound useful.
However, they don't seem to be documented at all. So once again, it is not at all clear if they are supposed to be an "official" API or not.
And if they are official, why are they only available for LaTeX?

**Comment by jfbu** [comment]
@mgeier regarding `stable` vs `master` it is documented at http://www.sphinx-doc.org/en/stable/devguide.html#contributing-to-sphinx:

> Fork the repository on Github to start making your changes to the master branch for next major version, or stable branch for next minor version.

**Comment by mgeier** [comment]
For the record, I've worked around the problem using `pushbody()` and `popbody()`: https://github.com/spatialaudio/nbsphinx/pull/85.

@jfbu Thanks for the pointer. I think the docs should elaborate a bit more on that. An occasional contributor like me still doesn't know if a suggested change is supposed to go to the next major or minor version.

Also, I'm suggesting #3252 for this page to be more visible to developers.
I've based that on `stable`, but I'm still not sure if that's right ...

**Comment by tk0miya** [comment]
>I should have probably based this PR on stable in the first place, but I never know which branch to base PRs on. I'm normally confused by the available branches.
>Is there some documentation that explains that?
>And what does stable even mean for Sphinx?

Sorry, I don't know that.
Here, https://github.com/sphinx-doc/sphinx/pull/3227#issuecomment-266330488 is basic branch policy AFAIK.

Anyway, sphinx developers would lead appropriate branch on reviewing.
So you don't mind which branch is better.

>However, they don't seem to be documented at all. So once again, it is not at all clear if they are supposed to be an "official" API or not.

Ah, you're right. I'd forgetten. Surely, they're not documented as you said. And, it's not official API.
So We don't promise the compatibility of these interfaces in future version.
But they are used for a long time, so they would not be changed, I believe.

>And if they are official, why are they only available for LaTeX?

In LaTeX writer, some bodies should be rewritten because notation rule has changed by context. For example, inside table, we have to use escaped return codes.
To do that, `pushbody()` and `popbody()` are used in LaTeX writer.

On the other hands, in other writers, rewriting bodies is not needed.

**Comment by mgeier** [comment]
> So you don't mind which branch is better.

OK, that's good to know. You should probably add something to the docs like "In case of doubt, base your commits on the `master` branch, the Sphinx core developers will then decide if the `stable` branch is more appropriate".

Do you want to "upgrade" `pushbody()` and `popbody()` to official API functions?
I think that would make sense.
However, I think they should be implemented for all writers, even if currently they are only used for LaTeX.

**Comment by QuLogic** [comment]
Please update the milestone since this apparently did not go into 1.5.1.

**Comment by tk0miya** [comment]
Updated. thanks,

**Pull request merged**
